Create metadata dump for a project

EddieLF opened the pull request:
Will leave this here for now, once we have the vcf files sorted out I will add some code to pull that out and move everything into the release bucket.

EddieLF committed:
added type hints

EddieLF committed:
removed shebang

MattWellie commented:
Happy to review this, just checking in if it's worth holding off on a metamist-heavy script until after the graphQL migration, or whether we take this for now, and use it as a template for future changes

cassimons commented:
Yea we need a dump done later this week so we may as well put this in as is then revisit post updates. Having said that, I did not think the existing endpoints are going to be retired, or did I miss that?

illusional commented:
They won't be retired. Should probably do this using GraphQL tbh. There is very basic graphql support already.

MattWellie commented on data_transfer/get_project_metadata.py:
```diff
@@ -0,0 +1,114 @@
+
+@click.command()
+@click.option('--project')
+@click.option('--output-path')
```
This `output_path` needs to exist already, i.e. this process will fail instead of making a new folder. It also won't function in GCP, so we're committing to running this locally and uploading the results. Not a problem, just an observation.

MattWellie commented on data_transfer/get_project_metadata.py:
```diff
@@ -0,0 +1,114 @@
+    individual_hpo_terms = get_hpo_terms(individual_metadata)
+
+    pedigrees = get_pedigrees(project)
+    print(type(pedigrees[0]))
```
?

MattWellie commented on data_transfer/get_project_metadata.py:
```diff
@@ -0,0 +1,114 @@
+    """Writes the HPO terms file, pedigree file, and sample map file to the output path"""
+
+    # HPO terms tsv
+    with open(f'{output_path}/hpo_terms.tsv', 'w') as f:
```
It's good practice to open files with a specific encoding, a more stringent linter would complain

```suggestion
    with open(f'{output_path}/hpo_terms.tsv', 'w', encoding='utf-8') as f:
```

MattWellie commented on data_transfer/get_project_metadata.py:
```diff
@@ -0,0 +1,114 @@
+    # Sample map
+    with open(f'{output_path}/external_translation.json', 'w') as f:
+        json.dump(sample_map, f, indent=4)
+    logging.info(f'Wrote sample ID map json to {output_path}.')
```
If you want to get spicy a nice last step might be compressing all the files into a single .zip, which would simplify subsequent transfers

EddieLF replied:
pure genius

MattWellie reviewed (approved):
Tried out locally, works great

EddieLF committed:
use graphql, added tar all metadata files and upload to release

EddieLF committed:
formatting fix

EddieLF committed:
sample_map dictionary fix

EddieLF commented:
@illusional Thanks for sussing this, the pedigree and sample mapping calls now use GraphQL queries. Any idea how I might get the individual metadata fields like HPO terms in the same way? They seem pretty wrapped up in the `participant_phenotype` table which I'm struggling to access through GraphQL.

@MattWellie Thanks for the comments, I've implemented your suggested changes. 
The `ruff`/`flake8-bandit` pre-commit hooks are flagging the uses of `subprocess.call()` for compressing the files and uploading them to the release bucket as a security issue, so I've added some `# noqa` ignores to those lines. Not sure if this is a good idea or if I should try another approach, its unclear if there's actually a security issue with these lines. I don't see how the `tar` or `gcloud` commands could be vulnerable to any kind of injection...

illusional commented:
@EddieLF ah right, they are not currently exposed, it's not super practical to address this in the current release version due to some much needed refactoring, but I will add this to graphql in the future.

MattWellie commented:
Instead of a direct call you could use `zipfile`, which I think is in the Python standard library

```python
import zipfile

# Create a ZipFile object
with zipfile.ZipFile('my_archive.zip', 'w') as my_zip:
    # Add files to the archive
    my_zip.write('file1.txt')
    my_zip.write('file2.txt')
    my_zip.write('file3.txt')
```

EddieLF committed:
Use zipfile and gcs upload not subprocesses

EddieLF commented:
@MattWellie thanks for the zip tip. I think this works fine now, so the last part will be adding in a step to copy over the files generated by the seqr load. 

Is this logical?

1. Get analysis records from the project where `AnalysisType == 'custom'`
2. Filter the results for the meta field `type: 'dataset-vcf'` and take the latest completed analysis
3. Copy the .vcf.bgz and index from the analysis object's output path into the release bucket directory, alongside the metadata.

MattWellie commented:
If this is all being done as part of one continuous process, I think the release bucket copy might be a bit of a pain. It might require authentication as a different user compared to the GCP authentication needed for metamist queries. AFAIK that's one of the reasons that step of the copy-to-release is done using single-purpose scripts.

I might be completely wrong about that, just spitballing. Maybe @illusional or @lgruen will know more.

lgruen commented:
With `full` permissions you should be able to write to the `release` bucket.

MattWellie commented:
I think I was confusing `release` and `upload`. Ignore me!

EddieLF committed:
vcf copier

EddieLF committed:
added vcf and index capturing and upload to release

EddieLF committed:
ignore linter for shebang

EddieLF commented:
@MattWellie I added a section to pick up the VCF and index that get generated by the seqr loader from your changes, is this premature since we don't have analyses like this yet?

MattWellie commented:
> is this premature since we don't have analyses like this yet?

Yeah, I reckon this will cause a cheeky little failure at the moment... I can't remember if the VCF extraction stage is merged in yet, might be worth running it

MattWellie commented:
@EddieLF a lil blast from the past here, but there's now at least one VCF generated and recorded in metamist 🎉 

Something worth mentioning is that there is a metamist entry added when a DatasetVCF job is queued, with the attributes:

```
status: queued
active: true
timestamp_completed: null
```

This entry would be updated when the process completes, but until then the query that you're phrasing [here](https://github.com/populationgenomics/rare-disease/pull/25/commits/0f43d2af3362702c60906257f19d519462ddbd37) would still pick up that db entry, even though the file doesn't exist yet, and the timestamp isn't set, so the script will crash. It also looks like if the analysis entry is queued and the job creating it fails, this queued entry will linger indefinitely...

Can you add `status` to the returned attributes, and filter for `status=="COMPLETED"` to make sure we're avoiding these?

I thought I could add something like this to the query to filter to these, but the response says that status isn't a top level gql attribute on analysis. @illusional Could we have that added? (Or will it be added once prod and dev sync up):

```
analyses(type: CUSTOM, active: true, status: COMPLETED)

> Unknown argument 'status' on field 'GraphQLProject.analyses'
```

illusional commented:
You'll find it on the sequencing-group change. Most entities allow a much greater degree of filtering, you can see how it will look on the [dev](https://sample-metadata-dev.populationgenomics.org.au/graphql) instance.

EddieLF committed:
better filtering for VCF analysis finder

EddieLF committed:
Remove unused imports

EddieLF commented:
@MattWellie thanks for flagging that. I've added in the required filtering for `status == 'COMPLETED'` and simplified the logic to find the most recent analysis to just use the `sorted()` function based on the timestamps.
